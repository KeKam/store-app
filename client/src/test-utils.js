import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';

import rootReducer from './redux/root-reducer';

// const render = (
//   component,
//   {
//     initialState,
//     store = createStore(rootReducer, initialState),
//     ...renderOptions
//   } = {}
// ) => {
//   return {
//     ...rtlRender(<Provider store={store}>{component}</Provider>),
//     store,
//   };
//   // const Wrapper = ({ children }) => {
//   //   return <Provider store={store}>{children}</Provider>;
//   // };
//   // return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
// };

const render = (component, { initialState } = {}) => {
  const actions = [];
  const observerMiddleware = () => (next) => (action) => {
    actions.push(action);
    return next(action);
  };
  const store = createStore(
    rootReducer,
    initialState,
    applyMiddleware(observerMiddleware)
  );
  const utils = {
    dispatch(action) {
      return store.dispatch(action);
    },
    getDispatchedActions() {
      return actions;
    },
    getState() {
      return store.getState();
    },
  };
  return {
    ...rtlRender(<Provider store={store}>{component}</Provider>),
    ...utils,
  };
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { render };
