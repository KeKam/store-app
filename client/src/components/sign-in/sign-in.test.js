import React from 'react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

import { render, screen } from '../../test-utils';
import {
  startEmailSignIn,
  startGoogleSignIn,
} from '../../redux/user/user.actions';
import SignIn from './sign-in';

describe('SignIn component', () => {
  it('should render SignIn component', () => {
    const { asFragment } = render(<SignIn />);

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText('I already have an account')).toBeInTheDocument();
    expect(
      screen.getByText('Sign in with your email and password')
    ).toBeInTheDocument();
    expect(screen.getByTestId('email-input')).toHaveDisplayValue('');
    expect(screen.getByTestId('password-input')).toHaveDisplayValue('');
    expect(screen.getByRole('button', { name: 'Sign in' })).toBeInTheDocument();
    expect(
      screen.getByRole('button', { name: 'Sign in with Google' })
    ).toBeInTheDocument();
  });

  it('should change email value on change', () => {
    render(<SignIn />);

    userEvent.type(screen.getByTestId('email-input'), 'testUser@gmail.com');

    expect(screen.getByTestId('email-input')).toHaveDisplayValue(
      'testUser@gmail.com'
    );
  });

  it('should change password value on change', () => {
    render(<SignIn />);

    userEvent.type(screen.getByTestId('password-input'), 'testuser123');

    expect(screen.getByTestId('password-input')).toHaveDisplayValue(
      'testuser123'
    );
  });

  it('should dispatch startEmailSignin on click', () => {
    const mockUserCredentials = {
      email: '',
      password: '',
    };
    const { getDispatchedActions } = render(<SignIn />);

    userEvent.click(screen.getByRole('button', { name: 'Sign in' }));

    expect(getDispatchedActions()).toContainEqual(
      startEmailSignIn(mockUserCredentials)
    );
  });

  it('should dispatch startGoogleSignin on click', () => {
    const { getDispatchedActions } = render(<SignIn />);

    userEvent.click(
      screen.getByRole('button', { name: 'Sign in with Google' })
    );

    expect(getDispatchedActions()).toContainEqual(startGoogleSignIn());
  });
});
