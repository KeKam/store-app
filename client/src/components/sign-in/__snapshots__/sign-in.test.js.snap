// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn component should render SignIn component 1`] = `
<DocumentFragment>
  <div
    class="sc-ifAKCX llvkyD"
  >
    <h2
      class="sc-EHOje cciXSq"
    >
      I already have an account
    </h2>
    <span>
      Sign in with your email and password
    </span>
    <form>
      <div
        class="sc-bdVaJa eEzrUY"
      >
        <input
          class="sc-bwzfXH eKtSpl"
          data-testid="email-input"
          name="email"
          required=""
          type="email"
          value=""
        />
        <label
          class="sc-htpNat gLmdTI"
          value=""
        >
          Email
        </label>
      </div>
      <div
        class="sc-bdVaJa eEzrUY"
      >
        <input
          class="sc-bwzfXH eKtSpl"
          data-testid="password-input"
          name="password"
          required=""
          type="password"
          value=""
        />
        <label
          class="sc-htpNat gLmdTI"
          value=""
        >
          Password
        </label>
      </div>
      <div
        class="sc-bZQynM iXkwVI"
      >
        <button
          class="sc-bxivhb jzXiqt"
          type="submit"
        >
           Sign in 
        </button>
        <button
          class="sc-bxivhb gxnkwT"
          type="button"
        >
          Sign in with Google
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
